system/service/auth.go

```diff
@@ -600,1082 +514,128 @@ func (svc auth) Impersonate(ctx context.Context, userID uint64) (u *types.User,
 
-		u, err = store.LookupUserByID(ctx, svc.store, userID)
-		if errors.IsNotFound(err) {
-			return AuthErrPasswordChangeFailedForUnknownUser(aam)
-		}
+	}
 
-		aam.setUser(u)
-		ctx = internalAuth.SetIdentityToContext(ctx, u)
+	if err = svc.LoadRoleMemberships(ctx, u); err != nil {
+		return
+	}
 
-		cc, _, err = store.SearchCredentials(ctx, svc.store, types.CredentialsFilter{
-			Kind:    credentialsTypePassword,
-			OwnerID: userID,
-			Deleted: filter.StateInclusive})
+	{
+		var eapSec *types.ExternalAuthProviderSecurity
 
-		if err != nil {
-			return err
-		}
+		switch p.Provider {
+		case credentialsTypePassword:
+			// nothing to do with password provider
 
-		if !svc.CheckPassword(oldPassword, true, cc) {
-			return AuthErrPasswordResetFailedOldPasswordCheckFailed(aam)
-		}
+		case "saml":
+			// we need to fetch SAML security settings from different part of settings
+			eapSec = &CurrentSettings.Auth.External.Saml.Security
 
-		if svc.CheckPassword(newPassword, false, cc) {
-			return AuthErrPasswordSetFailedReusedPasswordCheckFailed(aam)
-		}
+		default:
+			eap := CurrentSettings.Auth.External.Providers.FindByHandle(p.Provider)
 
-		if err != svc.SetPasswordCredentials(ctx, userID, newPassword) {
-			return err
-		}
+			if eap != nil {
+				eapSec = &eap.Security
+			}
 
-		if err = svc.RemoveAccessTokens(ctx, u); err != nil {
-			return err
 		}
 
-		return nil
-	}()
-
-	return svc.recordAction(ctx, aam, AuthActionChangePassword, err)
-}
-
-func (svc auth) hashPassword(password string) (hash []byte, err error) {
-	return bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
-}
-
-func (svc auth) CheckPasswordStrength(password string) bool {
-	pwdL := len(password)
-
-	// Ignore defined password constraints
-	if !svc.settings.Auth.Internal.PasswordConstraints.PasswordSecurity {
-		return true
-	}
-
-	// Check the password length
-	minL := math.Max(float64(passwordMinLength), float64(svc.settings.Auth.Internal.PasswordConstraints.MinLength))
-	if pwdL < int(minL) || pwdL > passwordMaxLength {
-		return false
-	}
-
-	// Check special constraints
-	// - numeric characters
-	count := svc.settings.Auth.Internal.PasswordConstraints.MinNumCount
-	if count > 0 {
-		rr := regexp.MustCompile("[0-9]")
-		if uint(len(rr.FindAllStringIndex(password, -1))) < count {
-			return false
+		if eapSec != nil {
+			// if authenticated with external auth provider
+			// there might be additional roles that need to be
+			// set to this security session
+			u.SetRoles(internalAuth.ApplyRoleSecurity(
+				payload.ParseUint64s(eapSec.PermittedRoles),
+				payload.ParseUint64s(eapSec.ProhibitedRoles),
+				payload.ParseUint64s(eapSec.ForcedRoles),
+				u.Roles()...,
+			)...)
 		}
 	}
 
-	// - special characters
-	count = svc.settings.Auth.Internal.PasswordConstraints.MinSpecialCount
-	if count > 0 {
-		rr := regexp.MustCompile("[^0-9a-zA-Z]")
-		if uint(len(rr.FindAllStringIndex(password, -1))) < count {
-			return false
+	if c != nil {
```

Comment:
can `c` ever be `nil`? I'd imagine proc login should never happen if there is no `c`